### ISSUE TITLE
Calling parallel process() for a very large file gets stuck

### ISSUE THREAD

**Issue by icui** [SeismicData/pyasdf#62]
I'm able to use the parallel process() function for medium sized files without problem (on Summit), but when I tried to process a 112GB file with 32,925 traces, the program will always get stuck. I've tried different node configurations (1 node 6 CPUs, 1 node 42 CPUs, 2 nodes 84 CPUs, etc.) but none of them worked. I was wondering if this related to #32 and is there any workaround to this problem?

Below is the full debug info.
Launching processing using MPI on 6 processors.
MASTER   received from WORKER 2 [WORKER_REQUESTS_ITEM] -- None
WORKER 2 sent to       MASTER   [WORKER_REQUESTS_ITEM] -- None
WORKER 1 sent to       MASTER   [WORKER_REQUESTS_ITEM] -- None
WORKER 5 sent to       MASTER   [WORKER_REQUESTS_ITEM] -- None
WORKER 4 sent to       MASTER   [WORKER_REQUESTS_ITEM] -- None
WORKER 3 sent to       MASTER   [WORKER_REQUESTS_ITEM] -- None
MASTER   sent to       WORKER 2 [MASTER_SENDS_ITEM] -- ('1A.CORRE', 'synthetic')
WORKER 2 received from MASTER   [MASTER_SENDS_ITEM] -- ('1A.CORRE', 'synthetic')
MASTER   received from WORKER 3 [WORKER_REQUESTS_ITEM] -- None
MASTER   sent to       WORKER 3 [MASTER_SENDS_ITEM] -- ('1A.NE00', 'synthetic')
MASTER   received from WORKER 4 [WORKER_REQUESTS_ITEM] -- None
MASTER   sent to       WORKER 4 [MASTER_SENDS_ITEM] -- ('1A.NE01', 'synthetic')
WORKER 3 received from MASTER   [MASTER_SENDS_ITEM] -- ('1A.NE00', 'synthetic')
RT Inventory created at 1995-06-29T12:22:32.070000Z
	Created by: SPECFEM3D_GLOBE/asdf-library
		    http://seismic-data.org
	Sending institution: SPECFEM3D_GLOBE
	Contains:
		Networks (1):
			1A
		Stations (1):
			1A.CORRE (N/A)
		Channels (3):
			1A.CORRE.S3.MXZ, 1A.CORRE.S3.MXN, 1A.CORRE.S3.MXE
WORKER 4 received from MASTER   [MASTER_SENDS_ITEM] -- ('1A.NE01', 'synthetic')
MASTER   received from WORKER 5 [WORKER_REQUESTS_ITEM] -- None
MASTER   sent to       WORKER 5 [MASTER_SENDS_ITEM] -- ('1A.NE02', 'synthetic')
WORKER 5 received from MASTER   [MASTER_SENDS_ITEM] -- ('1A.NE02', 'synthetic')
MASTER   received from WORKER 1 [WORKER_REQUESTS_ITEM] -- None
MASTER   sent to       WORKER 1 [MASTER_SENDS_ITEM] -- ('1A.NE03', 'synthetic')
WORKER 1 received from MASTER   [MASTER_SENDS_ITEM] -- ('1A.NE03', 'synthetic')
WORKER 2 sent to       MASTER   [WORKER_REQUESTS_ITEM] -- None
MASTER   received from WORKER 2 [WORKER_REQUESTS_ITEM] -- None
MASTER   sent to       WORKER 2 [MASTER_SENDS_ITEM] -- ('1A.NE04', 'synthetic')
WORKER 2 received from MASTER   [MASTER_SENDS_ITEM] -- ('1A.NE04', 'synthetic')

**Comment by krischer** [comment]
Is that the full output? Unfortunately it does not provide enough information for me to understand what is going on and at that scale I cannot easily reproduce it. It would be great if you could investigate the problem a bit and tell me what you find but I understand that this is not easy.

What should also work is to just disable MPI for `pyasdf` - this would fall back to using the `multiprocessing` module of Python. It should scale decently well to a full node and it uses a round robin method to organize actually writing to the file on disc.

To do this simply create the `ASDFDataSet` object with `pyasdf.ASDFDataSet(..., mpi=False)`. Then in your batch submission file (I don't know which queuing system Summit uses) make sure to submit the job in a way that it does not use `mpiexec`/`mpirun` (or if you use it use a single rank) but still make sure that it has access to a full node. `pyasdf` should then just take all cores of a node and process the data with it. **This will not work for more than one node!**

**Comment by icui** [comment]
Yes this is the full output. I'll try to investigate more on this, thanks.

**Comment by krischer** [comment]
That would be great - thanks!

I don't really understand why the inventory is printed in the above output, otherwise - if sorted by worker - it is fine (in the current `pyasdf` implementation rank 0 is always the `MASTER` and all workers [all other ranks] can request new stations to process:

### Worker 1
```
WORKER 1 sent to MASTER [WORKER_REQUESTS_ITEM] -- None
MASTER received from WORKER 1 [WORKER_REQUESTS_ITEM] -- None
MASTER sent to WORKER 1 [MASTER_SENDS_ITEM] -- ('1A.NE03', 'synthetic')
WORKER 1 received from MASTER [MASTER_SENDS_ITEM] -- ('1A.NE03', 'synthetic')
```
### Worker 2
```
MASTER received from WORKER 2 [WORKER_REQUESTS_ITEM] -- None
WORKER 2 sent to MASTER [WORKER_REQUESTS_ITEM] -- None
MASTER sent to WORKER 2 [MASTER_SENDS_ITEM] -- ('1A.CORRE', 'synthetic')
WORKER 2 received from MASTER [MASTER_SENDS_ITEM] -- ('1A.CORRE', 'synthetic')
```
---
```
WORKER 2 sent to MASTER [WORKER_REQUESTS_ITEM] -- None
MASTER received from WORKER 2 [WORKER_REQUESTS_ITEM] -- None
MASTER sent to WORKER 2 [MASTER_SENDS_ITEM] -- ('1A.NE04', 'synthetic')
WORKER 2 received from MASTER [MASTER_SENDS_ITEM] -- ('1A.NE04', 'synthetic')
```
### Worker 3
```
WORKER 3 sent to MASTER [WORKER_REQUESTS_ITEM] -- None
MASTER received from WORKER 3 [WORKER_REQUESTS_ITEM] -- None
MASTER sent to WORKER 3 [MASTER_SENDS_ITEM] -- ('1A.NE00', 'synthetic')
WORKER 3 received from MASTER [MASTER_SENDS_ITEM] -- ('1A.NE00', 'synthetic')
```
### Worker 4
```
WORKER 4 sent to MASTER [WORKER_REQUESTS_ITEM] -- None
MASTER received from WORKER 4 [WORKER_REQUESTS_ITEM] -- None
MASTER sent to WORKER 4 [MASTER_SENDS_ITEM] -- ('1A.NE01', 'synthetic')
WORKER 4 received from MASTER [MASTER_SENDS_ITEM] -- ('1A.NE01', 'synthetic')
```
### Worker 5
```
WORKER 5 sent to MASTER [WORKER_REQUESTS_ITEM] -- None
MASTER received from WORKER 5 [WORKER_REQUESTS_ITEM] -- None
MASTER sent to WORKER 5 [MASTER_SENDS_ITEM] -- ('1A.NE02', 'synthetic')
WORKER 5 received from MASTER [MASTER_SENDS_ITEM] -- ('1A.NE02', 'synthetic')
```

**Comment by icui** [comment]
Hi Lion, I found that the process() function will stuck when calling station_group.copy() during the creation of the new dataset (line 2111). When I remove station_group.copy(), the function will work fine. Do you have any idea what the problem is?

**Comment by krischer** [comment]
Can you try replacing 

```python
station_group.copy(
    source=data, dest=group, name="StationXML"
)
```

with 

```python
group["StationXML"] = data[:]
```

This might be a bit less efficient but I doubt it matters and if it works it is the better solution. Also make sure to check the StationXML files in the result in case it work.

**Comment by icui** [comment]
> Can you try replacing
> 
> ```python
> station_group.copy(
>     source=data, dest=group, name="StationXML"
> )
> ```
> 
> with
> 
> ```python
> group["StationXML"] = data[:]
> ```
> 
> This might be a bit less efficient but I doubt it matters and if it works it is the better solution. Also make sure to check the StationXML files in the result in case it work.

Yes this is working. The StationXML in the result also seems fine.

**Comment by krischer** [comment]
Perfect :-)

Fixed in 08419cf743ad8a0d61f0599919f8e6ea9650486c and I also just released 0.7.3 that contains the fix. It is already out on pypi and will be on conda-forge in a bit.

Thanks for your help with this!
